fontir/src/variations.rs

```diff
@@ -113,6 +121,75 @@ impl VariationModel {
     pub fn default_location(&self) -> &NormalizedLocation {
         &self.default
     }
+
+    /// Convert absolute positions at master locations to offsets.
+    ///
+    /// <ul>
+    ///     <li>All keys in points must be known to the variation model.</li>
+    ///     <li>All point vectors must have the same length.</li>
+    ///     <li>A point vector for the default location is required</li>
+    /// </ul>
+    ///
+    /// Note that it is NOT required to provide a point sequence for every location known
+    /// to the variation model.
+    ///
+    /// Returns a delta, in the form of a [Vec2], for every input point. Intended use is to support
+    /// construction of a variation store.
+    ///
+    /// Rust version of <https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/varLib/models.py#L449-L461>
+    pub fn deltas(
+        &self,
+        point_seqs: HashMap<NormalizedLocation, Vec<Point>>,
+    ) -> Result<HashMap<NormalizedLocation, Vec<Vec2>>, DeltaError> {
+        let Some(defaults) = point_seqs.get(&self.default) else {
+            return Err(DeltaError::DefaultUndefined);
+        };
+        if point_seqs.values().any(|pts| pts.len() != defaults.len()) {
+            return Err(DeltaError::InconsistentNumbersOfPoints);
+        }
+
+        let mut result: HashMap<NormalizedLocation, Vec<Vec2>> = HashMap::new();
+        // self.locations is sorted such that[i] is only influenced by[i+1..N]
+        // so we know subsequent spins won't invalidate our delta if we go in the same order
+        for (loc_idx, points) in self
+            .locations
+            .iter()
+            .filter_map(|loc| point_seqs.get(loc))
+            .enumerate()
+        {
+            let master_influences = &self.delta_weights[loc_idx];
+
+            let mut deltas = Vec::new();
+            for (idx, point) in points.iter().enumerate() {
+                deltas.push(
+                    // Find other masters that are active (have influence)
+                    master_influences
+                        .iter()
+                        .map(|(loc_idx, w)| (&self.locations[*loc_idx], w.into_inner()))
+                        .filter_map(|(loc, w)| result.get(loc).map(|deltas| (deltas, w)))
+                        .filter_map(|(deltas, w)| deltas.get(idx).map(|delta| (delta, w)))
+                        // Start with a vector to our destination, then subtract away influence from other active masters
+                        .fold(point.to_vec2(), |acc, (other, other_weight)| {
```

Comment:
wow, I'm lost in the chain of maps...

Reply:
Will add comments or go back to a for-loop.

Reply:
Better?

Reply:
yes, thanks! :)